text/unify-theorem-definition-fixpoint-syntax.md

```diff
@@ -0,0 +1,193 @@
+  between a `Theorem` and a `Fixpoint`.  Still, there is no reason why
+  we should explicitly distinguish between a `Definition` and a
+  `Fixpoint` but not between a non-recursive and a recursive
+  `Theorem`.
+
+  The proposed solution is to keep the existing syntax `Theorem
+  ... with` but encourage that it comes with an explicit `recursive`
+  annotation (in the form of an attribute).  This also opens the way
+  to supporting (non-mutually) recursive `#[ recursive ] Theorem`, and
+  `#[ recursive ] Definition` as an equivalent to `Fixpoint` to define
+  (possibly mutually) recursive objects.
+
+# Detailed design
+
+The current syntax is:
+
+- for `Definition` / `Example`:
+  ```
+  {| Definition | Example } @ident_decl @def_body
+
+  def_body ::= {* @binder } {? : @type } := {? @reduce } @term
+             | {* @binder } : @type
+  ```
+
+- for `Fixpoint`:
+  ```
+  Fixpoint @fix_def {* with @fix_def }
+
+  fix_def ::= @ident_decl {* @binder } {? @fixannot } {? : @type } {? := @term } {? @decl_notations }
+  ```
+
+- for `CoFixpoint`:
+  ```
+  CoFixpoint @cofix_def {* with @cofix_def }
+
+  cofix_def ::= @ident_decl {* @binder } {? : @type } {? := @term } {? @decl_notations }
+  ```
+
+- for `Theorem` and variants:
+  ```
+  @thm_token @ident_decl {* @binder } : @type {* with @ident_decl {* @binder } : @type }
+
+  thm_token ::= Theorem
+              | Lemma
+              | Fact
+              | Remark
+              | Corollary
+              | Proposition
+              | Property
+  ```
+
+The new syntax would be:
+
+```
+@decl_token @decl_def {* with @decl_def }
+
+decl_token := Definition
+            | Example
+            | Fixpoint
+            | CoFixpoint
+            | Theorem
+            | Lemma
+            | Fact
+            | Remark
+            | Corollary
+            | Proposition
+            | Property
+
+decl_def := @ident_decl {* @binder } {? @fixannot } @decl_body {? @decl_notations }
+
+decl_body := {? : @type } := {? @reduce } @term
+           | : @type
+```
+
+with the following newly supported attributes:
+
+- `recursive` and `corecursive` to declare a fixpoint / cofixpoint:
+
+  - attribute `recursive` is implicit for `Fixpoint` and rejected by
+    `CoFixpoint`;
+  - attribute `corecursive` is implicit for `CoFixpoint` and rejected
+    by `Fixpoint`;
+  - attribute `recursive` or `corecursive` is compulsory to use the
+    `with` variant of `Definition`;
+  - using the `with` variant of `Theorem` without attribute
+    `recursive` or `corecursive` raises a deprecation message and is
+    interpreted as if `recursive` had been used;
+  - the `@fixannot` clause is reserved to the `recursive` attribute
+    being used (or implicit in the case of `Fixpoint`).
+
+- `opaque` and `transparent` to declare an object opaque or
+  transparent like `Qed` and `Defined` do (note that this is not the
+  same as what the `Opaque` and `Transparent` commands do):
```

Comment:
This is the question I'm asking in the "Unresolved questions" section.

Reply:
IMVHO, I think `qed` and `defined` would be strictly better names in the current scenario.
Otherwise, please pick *fresh* words that can be used consistently, and make sense if coq/coq#3389 is accepted.

Ideally, technical terms and their meanings would be in a one-to-one mapping, absent strong reasons for a specific case.

Reply:
FTR at the coq call I heard `seal` and `unseal`, for flipping the "Qed opaque" status. We could then keen `opacity(number)` for the "other opacity".